lib/compress.js

```diff
@@ -1980,12 +1980,14 @@ merge(Compressor.prototype, {
         if (!compressor.option("side_effects")) return false;
         if (this.pure !== undefined) return this.pure;
         var pure = false;
-        var comments, last_comment;
+        var comments, pure_comment;
         if (this.start
             && (comments = this.start.comments_before)
             && comments.length
-            && /[@#]__PURE__/.test((last_comment = comments[comments.length - 1]).value)) {
-            pure = last_comment;
+            && (pure_comment = find_if(function (comment) {
+                return /[@#]__PURE__/.test(comment.value);
+            }, comments))) {
+            pure = pure_comment;
```

Comment:
@alexlamsl Are you okay with the efficiency of `find_if` in this scenario? If so, LGTM.

Reply:
`pure` & `pure_comment` now feels redundant as `find_if()` will return `undefined`, e.g. falsy, if no comment has matched.

Reply:
> pure & pure_comment now feels redundant as find_if() will return undefined, e.g. falsy, if no comment has matched.

It's not redundant. It would screw up the caching of `return this.pure = pure;`. `undefined` is not falsey in this scenario - it means "not set". `false` means no pure annotation found previously and no need to scan again.

Reply:
Ah okay - thanks for the clarification :+1:

Reply:
@alexlamsl I see your point. Could get rid of `var pure` and just:
```
return this.pure = !!pure_comment;
```

Reply:
Sorry - my last remark wouldn't work. Need to set `this.pure` to `false` or to the comment:
```
            if (this.pure) {
                compressor.warn("Dropping __PURE__ call [{file}:{line},{col}]", this.start);
                this.pure.value = this.pure.value.replace(/[@#]__PURE__/g, ' ');
            }
```

Reply:
I think this should work:
```patch
-        return this.pure = pure;
+        return this.pure = pure_comment || false;
```

Reply:
its already initialized with `false` [few line above](https://github.com/mishoo/UglifyJS2/pull/2305/files/fb8b222002f3bafc9cac35618ad1f9a21abe5bf0#diff-230c53e13e3ee367aec7d49470eb42caR1982)

Reply:
`pure_comment` is not initialized to `false`:

https://github.com/mishoo/UglifyJS2/pull/2305/files#diff-230c53e13e3ee367aec7d49470eb42caR1983

and [`find_if`](https://github.com/mishoo/UglifyJS2/blob/17436218891781a5dc4fad8a9c23080c254aa695/lib/utils.js#L58-L63) returns `undefined` if nothing found.

Reply:
> Assuming we want to eliminate `var pure` ...

Or just leave PR as is. It works.

Reply:
Yeah, don't sweat too much over my OCD... :sweat_smile: